sphinx/domains/std.py

```diff
@@ -170,15 +172,42 @@ def handle_signature(self, sig: str, signode: desc_signature) -> str:
                                location=signode)
                 continue
             optname, args = m.groups()
-            if optname.endswith('[') and args.endswith(']'):
+            if optname[-1] == '[' and args[-1] == ']':
                 # optional value surrounded by brackets (ex. foo[=bar])
                 optname = optname[:-1]
                 args = '[' + args
 
             if count:
-                signode += addnodes.desc_addname(', ', ', ')
+                if self.env.config.option_detailed_parse:
+                    signode += addnodes.desc_sig_punctuation(',', ',')
+                    signode += addnodes.desc_sig_space()
+                else:
+                    signode += addnodes.desc_addname(', ', ', ')
             signode += addnodes.desc_name(optname, optname)
-            signode += addnodes.desc_addname(args, args)
+            if self.env.config.option_detailed_parse:
+                add_end_bracket = False
+                if not args:
+                    continue
+                if args[0] == '[' and args[-1] == ']':
+                    add_end_bracket = True
+                    signode += addnodes.desc_sig_punctuation('[', '[')
+                    args = args[1:-1]
+                if args[0] == ' ':
+                    signode += addnodes.desc_sig_space()
+                    args = args.strip()
+                if args[0] == '=':
+                    signode += addnodes.desc_sig_punctuation('=', '=')
+                    args = args[1:]
+                for part in samp_role.parse(args):
+                    if isinstance(part, nodes.Text):
+                        text = part.astext()
+                        signode += nodes.Text(text, text)
```

Comment:
Avoid a DeprecationWarning
```suggestion
                        signode += nodes.Text(part.astext())
```